samples: net: add put and post methods to coap

tarunkum opened the pull request:
In sample application of COAP put and post
    methods are implemented in coap-server.c
    and coap-client.c

Signed-off-by: Tarun Kumar Agarwal tarun.kumar.agarwal@intel.com

jukkar commented on samples/net/coap_client/src/coap-client.c:
```diff
@@ -42,14 +42,23 @@ static struct net_mgmt_event_callback cb;
+
+	printk("\n");
+
+	for (i = 0; i < len; i++)
```
Please add {} even for one liners. The original code did not had these but if you make changes then it is worth adding those braces.

jukkar commented on samples/net/coap_client/src/coap-client.c:
```diff
@@ -162,8 +171,39 @@ static void retransmit_request(struct k_work *work)
+						.sin6_family = AF_INET6 };
+
+	r = net_context_get(PF_INET6, SOCK_DGRAM, IPPROTO_UDP, &context);
+		if (r) {
```
Indentation is wrong either above or below, so please fix accordingly.

jukkar commented on samples/net/coap_client/src/coap-client.c:
```diff
@@ -174,28 +214,17 @@ static void event_iface_up(struct net_mgmt_event_callback *cb,
-		printk("Could not bind the context\n");
-		return;
-	}
+	u8_t put_payload[40] = "ZEPHYR COAP put method under test";
```
Please do not put length when defining the array of this kind, just let the compiler calculate the length for you.

jukkar commented on samples/net/coap_client/src/coap-client.c:
```diff
@@ -174,28 +214,17 @@ static void event_iface_up(struct net_mgmt_event_callback *cb,
-		printk("Could not get an UDP context\n");
-		return;
-	}
+	u8_t My_COAP_token1[8] = "puttok";
```
Is there a reason why the length is specified here, for example token1 is only 7 chars long. Could you omit the length?

jukkar commented on samples/net/coap_client/src/coap-client.c:
```diff
@@ -211,9 +240,29 @@ static void event_iface_up(struct net_mgmt_event_callback *cb,
-			     8, coap_next_token(),
-			     COAP_METHOD_GET, coap_next_id());
+	switch(coap_method)
+	{
```
Put the curly brace to same line as switch statement.

jukkar commented on samples/net/coap_client/src/coap-client.c:
```diff
@@ -211,9 +240,29 @@ static void event_iface_up(struct net_mgmt_event_callback *cb,
+		break;
+
+		default:
+		return;
```
Return indentation is wrong

jukkar commented on samples/net/coap_client/src/coap-client.c:
```diff
@@ -211,9 +240,29 @@ static void event_iface_up(struct net_mgmt_event_callback *cb,
+
+		case 1:
+			r = coap_packet_init(&request, pkt, 1, COAP_TYPE_CON,
+					7,My_COAP_token1, COAP_METHOD_PUT, coap_next_id());
```
Fix indentation. Also there should be space after ",". You could check these issues before submitting the patch, there is scripts/checkpatch.pl script just for this purpose.

rbtchc replied:
You can call coap_next_token to get an unique token w/o declaring My_COAP_token1/2

jukkar commented on samples/net/coap_client/src/coap-client.c:
```diff
@@ -211,9 +240,29 @@ static void event_iface_up(struct net_mgmt_event_callback *cb,
+
+		case 2:
+			r = coap_packet_init(&request, pkt, 1, COAP_TYPE_CON,7,
+					My_COAP_token2, COAP_METHOD_POST, coap_next_id());
```
Fix indentation

jukkar commented on samples/net/coap_client/src/coap-client.c:
```diff
@@ -264,6 +314,48 @@ static void event_iface_up(struct net_mgmt_event_callback *cb,
 
+
+	switch(coap_method)
+	{
```
{ in wrong line

jukkar commented on samples/net/coap_client/src/coap-client.c:
```diff
@@ -272,16 +364,18 @@ static void event_iface_up(struct net_mgmt_event_callback *cb,
 
 void main(void)
 {
+	//test coap get method
```
No c++ comments

jukkar commented on samples/net/coap_client/src/coap-client.c:
```diff
@@ -272,16 +364,18 @@ static void event_iface_up(struct net_mgmt_event_callback *cb,
-#if defined(CONFIG_NET_MGMT_EVENT)
+	event_iface_init();
+
+#if 0
```
Just remove dead code if it is not used.

jukkar reviewed (changes requested):
Please use scripts/checkpatch.pl to check your patch before submitting again, your patch has lot of style issues.

Also your Signed-off line is wrong, it should be your real email address. See example syntax in git log.

rbtchc commented on samples/net/coap_server/src/coap-server.c:
```diff
@@ -131,6 +131,41 @@ static int well_known_core_get(struct coap_resource *resource,
+	printk("\n %u bytes \n \n", len);
+}
+
+static void paload_dump(const char *s, u8_t *data, unsigned len,unsigned payloadlen)
```
paload -> payload

rbtchc commented on samples/net/coap_server/src/coap-server.c:
```diff
@@ -131,6 +131,41 @@ static int well_known_core_get(struct coap_resource *resource,
+
+static void paload_dump(const char *s, u8_t *data, unsigned len,unsigned payloadlen)
+{
+	unsigned i,LoopCount;
```
I believe Zephyr is not using CamelCase naming convention.
Alter "LoopCount" to loopcount or loop_count

rbtchc commented on samples/net/coap_server/src/coap-server.c:
```diff
@@ -131,6 +131,41 @@ static int well_known_core_get(struct coap_resource *resource,
+	LoopCount = len -payloadlen;
+
+	printk(" paload_dump =  %s \n ", s);
+	for (i = LoopCount; i < len; i++)
```
for (...) {
   ...
}

rbtchc commented on samples/net/coap_server/src/coap-server.c:
```diff
@@ -131,6 +131,41 @@ static int well_known_core_get(struct coap_resource *resource,
+
+	printk("\n");
+
+	for (i = LoopCount; i < len; i++)
```
same here

rbtchc commented on samples/net/coap_client/src/coap-client.c:
```diff
@@ -162,8 +171,39 @@ static void retransmit_request(struct k_work *work)
+
 static void event_iface_up(struct net_mgmt_event_callback *cb,
-			   u32_t mgmt_event, struct net_if *iface)
+			   u32_t mgmt_event, struct net_if *iface,u8_t coap_method)
```
use type "enum coap_method" here.

rbtchc commented on samples/net/coap_client/src/coap-client.c:
```diff
@@ -211,9 +240,29 @@ static void event_iface_up(struct net_mgmt_event_callback *cb,
-			     COAP_METHOD_GET, coap_next_id());
+	switch(coap_method)
+	{
+		case 0:
```
If we use type "enum coap_method" in the function interface, then we could use "COAP_METHOD__XXX" here as case id.

tarunkum commented:
Hello Jukka/Robert,

Thanks a lot for review. I have done the changes as per the given comments. I have uploaded the new changes in the same branch.

Regards,
Tarun

tarunkum commented:
Hello @jukkar / @rbtchc ,

I have removed all the warnings.

rbtchc commented:
@tarunkum you still need to fix sanity test failures. Following the linked posted by Shippable at the bottom of the PR. Also, there are still many code styling issues, you might want to fix those as well.

tarunkum commented:
@rbtchc , Looking into will try to solve all warnings as soon as possible.

tarunkum commented:
Hello @jukkar / @rbtchc ,

All the issues has resolved now. Please let me know if any changes are required.

rbtchc commented on samples/net/coap_client/src/coap-client.c:
```diff
@@ -196,6 +199,26 @@ static void event_iface_up(struct net_mgmt_event_callback *cb,
 	k_delayed_work_init(&retransmit_work, retransmit_request);
+}
+
+static void event_iface_up(struct net_mgmt_event_callback *cb,
```
this is no longer a callback function registered and called when an interface is up. please update the interface to remove the unnecessary parameters and change the name to avoid misleading

tarunkum replied:
I have to keep break this function into two then only I can test other methods at a same time. Otherwise it is not possible.

	event_iface_init();
	printk("\n ************ COAP Client  test 1 *************\n");
	event_iface_up(NULL, NET_EVENT_IF_UP, iface, COAP_METHOD_GET);

So that I have implemented. if you want to give new name to event_iface_init(); function let me know.

rbtchc commented on samples/net/coap_client/src/coap-client.c:
```diff
@@ -196,6 +199,26 @@ static void event_iface_up(struct net_mgmt_event_callback *cb,
+}
+
+static void event_iface_up(struct net_mgmt_event_callback *cb,
+			   u32_t mgmt_event, struct net_if *iface, u8_t code)
```
type "u8_t" -> "enum coap_method"
and perhaps we should call it "method" instead of "code"?

tarunkum replied:
done

rbtchc commented on samples/net/coap_client/src/coap-client.c:
```diff
@@ -196,6 +199,26 @@ static void event_iface_up(struct net_mgmt_event_callback *cb,
+	struct net_buf *frag;
+	int r;
+
+	u8_t put_payload[] = "ZEPHYR COAP put method test done";
```
is it necessary to have different payload contents for different requests?

rbtchc commented on samples/net/coap_client/src/coap-client.c:
```diff
@@ -235,7 +275,7 @@ static void event_iface_up(struct net_mgmt_event_callback *cb,
 	}
 
-	/* setup appdatalen */
+	/* setup app data length */
```
restore this line

tarunkum replied:
done

rbtchc commented on samples/net/coap_client/src/coap-client.c:
```diff
@@ -245,6 +285,7 @@ static void event_iface_up(struct net_mgmt_event_callback *cb,
 		return;
 	}
 
+
```
remove this line

tarunkum replied:
done

rbtchc commented on samples/net/coap_client/src/coap-client.c:
```diff
@@ -264,6 +305,45 @@ static void event_iface_up(struct net_mgmt_event_callback *cb,
 	/* Increase packet ref count to avoid being unref after sendto() */
 	coap_pending_cycle(pending);
 
+
```
remove this line

tarunkum replied:
done

rbtchc commented on samples/net/coap_client/src/coap-client.c:
```diff
@@ -211,9 +234,26 @@ static void event_iface_up(struct net_mgmt_event_callback *cb,
-	r = coap_packet_init(&request, pkt, 1, COAP_TYPE_CON,
-			     8, coap_next_token(),
-			     COAP_METHOD_GET, coap_next_id());
+	switch (code) {
```
line 237 ~ 256 can be simplified as
```

if (code == COAP_METHOD_DELETE) {
        printk("Delete method is not supported");
        return;
}

r = coap_packet_init(&request, pkt, 1, COAP_TYPE_CON,
                                  8, coap_next_token(), code,
                                  coap_next_id());
```

tarunkum replied:
Added the case for COAP_METHOD_DELETE method.

rbtchc commented on samples/net/coap_client/src/coap-client.c:
```diff
@@ -264,6 +305,45 @@ static void event_iface_up(struct net_mgmt_event_callback *cb,
 	coap_pending_cycle(pending);
 
+
+	switch (code) {
```
if we consolidate two payloads variable into one then we can replace the following as
`
if (code == COAP_METHOD_PUT || code == CODE_METHOD_POST) {
		r = coap_packet_append_payload_marker(&request);
		if (r) {
			net_pkt_unref(pkt);
			printk("Unable to append payload marker\n");
		}

		r = coap_packet_append_payload(&request, (u8_t *)put_payload,
				                                       strlen(put_payload));
		if (r < 0) {
			net_pkt_unref(pkt);
			printk("Not able to append payload\n");
		}
}
`

tarunkum replied:
I want to use different payload message for both methods. So that I have separated out.

rbtchc commented on samples/net/coap_client/src/coap-client.c:
```diff
@@ -272,24 +352,28 @@ static void event_iface_up(struct net_mgmt_event_callback *cb,
 		coap_pending_clear(pending);
 		return;
 	}
-
```
restore line 275 and line 277

rbtchc commented on samples/net/coap_server/src/coap-server.c:
```diff
@@ -131,6 +131,27 @@ static int well_known_core_get(struct coap_resource *resource,
+	printk("\npayloadlen = %u bytes\n", payloadlen);
+}
+
+
```
remove this line. only one blank line is required between function declarations.

tarunkum commented on samples/net/coap_client/src/coap-client.c:
```diff
@@ -162,19 +171,13 @@ static void retransmit_request(struct k_work *work)
 
-static void event_iface_up(struct net_mgmt_event_callback *cb,
-			   u32_t mgmt_event, struct net_if *iface)
+static struct net_context *context;
```
done

tarunkum commented on samples/net/coap_client/src/coap-client.c:
```diff
@@ -30,26 +31,34 @@ static const struct sockaddr_in6 mcast_addr = {
 	.sin6_family = AF_INET6,
 	.sin6_port = htons(MY_COAP_PORT)};
 
-static struct net_context *context;
```
done

tarunkum commented on samples/net/coap_client/src/coap-client.c:
```diff
@@ -19,6 +19,7 @@
 
 #define MY_COAP_PORT 5683
 
+
```
done

tarunkum commented on samples/net/coap_client/src/coap-client.c:
```diff
@@ -30,26 +31,34 @@ static const struct sockaddr_in6 mcast_addr = {
-
 static const char * const test_path[] = { "test", NULL };
 
+volatile u8_t test_complete;
```
Further implemented with Semaphore now.

tarunkum commented on samples/net/coap_client/src/coap-client.c:
```diff
@@ -30,26 +31,34 @@ static const struct sockaddr_in6 mcast_addr = {
+
+	printk("\nbytes = %u\n", len);
+
+	test_complete = 1;
```
Removed the test variable in msg_dump function

tarunkum committed:
samples: net: add put and post methods to coap

In sample application of COAP put and post
methods are implemented in coap-server.c
and coap-client.c

Signed-off-by: Tarun Kumar Agarwal <tarun.kumar.agarwal@intel.com>

tarunkum commented:
Hello @rbtchc ,

All the issues has resolved now. Please let me know if any further changes are required.

tarunkum commented:
@rbtchc I am waiting for this patch to be merged. Please let me know if any other change is required in code.

rbtchc commented on samples/net/coap_client/src/coap-client.c:
```diff
@@ -196,6 +196,22 @@ static void event_iface_up(struct net_mgmt_event_callback *cb,
 	k_delayed_work_init(&retransmit_work, retransmit_request);
+}
+
+static void event_iface_up(struct net_if *iface, u8_t method)
```
Suggest to change the function name to something like: send_coap_request(...)
And remove "struct net_if *iface" for you are not using it.

rbtchc commented on samples/net/coap_client/src/coap-client.c:
```diff
@@ -36,20 +36,27 @@ struct coap_pending pendings[NUM_PENDINGS];
 static const char * const test_path[] = { "test", NULL };
 
+/* define semaphores */
+K_SEM_DEFINE(COAP_sem, 0, 1);	/* starts off "available" */
```
I suppose this should be named in lower and do not append comments after the code.

rbtchc reviewed (changes requested):
I suppose you should ask @jukkar or @rveerama1 to review the patch as well for I'm not the code owner.

rveerama1 commented:
I am fine if you address @rbtchc comments.

tarunkum commented on samples/net/coap_client/src/coap-client.c:
```diff
@@ -211,9 +227,29 @@ static void event_iface_up(struct net_mgmt_event_callback *cb,
-			     8, coap_next_token(),
-			     COAP_METHOD_GET, coap_next_id());
+	switch (method) {
+	case COAP_METHOD_GET:
```
Modified the code for this comment.

tarunkum commented:
@rveerama1 I have addressed all the review comments. Please look into once.

tarunkum commented:
@rbtchc I have addressed all the review comments. Please look into once.

rbtchc commented on samples/net/coap_client/src/coap-client.c:
```diff
@@ -261,8 +288,43 @@ static void event_iface_up(struct net_mgmt_event_callback *cb,
+		r = coap_packet_append_payload_marker(&request);
+		if (r) {
+			net_pkt_unref(pkt);
+			printk("Unable to append payload marker\n");
```
Shouldn't we abort the sending request when payload append fail?

rbtchc commented on samples/net/coap_client/src/coap-client.c:
```diff
@@ -261,8 +288,43 @@ static void event_iface_up(struct net_mgmt_event_callback *cb,
+		r = coap_packet_append_payload(&request, (u8_t *)put_payload,
+				strlen(put_payload));
+		if (r < 0) {
+			net_pkt_unref(pkt);
```
Shouldn't we abort the sending request when payload append fail?

rbtchc commented on samples/net/coap_server/src/coap-server.c:
```diff
@@ -250,9 +289,18 @@ static int test_post(struct coap_resource *resource,
 	NET_INFO("*******\n");
 	NET_INFO("type: %u code %u id %u\n", type, code, id);
 	NET_INFO("*******\n");
+	payload_dump("post_paylod", payloadfrag->data, payloadfrag->len, len);
```
paylod -> payload

rbtchc commented on samples/net/coap_server/src/coap-server.c:
```diff
@@ -197,9 +222,18 @@ static int test_put(struct coap_resource *resource,
 	NET_INFO("*******\n");
 	NET_INFO("type: %u code %u id %u\n", type, code, id);
 	NET_INFO("*******\n");
+	payload_dump("put_paylod", payloadfrag->data, payloadfrag->len, len);
```
paylod -> payload

rbtchc commented on samples/net/coap_server/prj.conf:
```diff
@@ -10,3 +10,4 @@ CONFIG_COAP=y
 CONFIG_NET_APP_SETTINGS=y
 CONFIG_NET_APP_MY_IPV6_ADDR="2001:db8::1"
 CONFIG_NET_IF_MCAST_IPV6_ADDR_COUNT=5
+CONFIG_COAP_TEST_API_ENABLE=y
```
is this necessary?

tarunkum commented on samples/net/coap_client/src/coap-client.c:
```diff
@@ -261,8 +288,43 @@ static void event_iface_up(struct net_mgmt_event_callback *cb,
 	coap_reply_init(reply, &request);
 	reply->reply = resource_reply_cb;
 
-	/* Increase packet ref count to avoid being unref after sendto() */
```
@rbtchc Thanks I have addressed in new code changes. Please review it once.

rbtchc replied:
@tarunkum where?

tarunkum commented:
@rbtchc Please review the code. I have addressed all the review comments.

tarunkum commented:
@rbtchc I have merged the requested changes. Please review once.

rbtchc reviewed (commented):
LGTM

tarunkum commented:
@rveerama1 

Robert has reviewed & validated.

Please review one & let me know if any changes are required.

tarunkum commented:
@kumarvikash1 solved the merge conflict. Rebased the code.
